File: haskell/2023/1.pl
main :-
    open("1.txt", read, In),
    read_string(In, _, S), split_string(S, "\n", "\n", Lines),
    member(Part, [1,2]),
    maplist(process(Part), Lines, Xs), sum_list(Xs, Sum), writeln(Sum),
    fail.
main.

process(Part, S, N) :-
    string_codes(S, C),
    findall(X, (append(_, P, C), (numeric1(P, X) ; Part=2,numeric2(P,X))), Xs),
    Xs=[A|_], last(Xs, B), N is A*10+B.

numeric1([C|_], N) :- number_string(N, [C]).
numeric2(C, N) :- between(1,9,N), numbers(Numbers), nth1(N, Numbers, P), prefix(P, C).
numbers([`one`,`two`,`three`,`four`,`five`,`six`,`seven`,`eight`,`nine`]).

File: haskell/2023/1.hs
import Control.Arrow
import Data.Char
import Data.List
import Data.Maybe

main = do
  xs <- lines <$> readFile "1.txt"
  let g = uncurry (+) . first (*10) . (head &&& last) . mapMaybe numeric . tails
  mapM_ (print . sum . map g) [filter isDigit <$> xs, xs]
numeric (c:_) | isDigit c = Just (digitToInt c)
numeric s = find (\n -> (numbers !! n) `isPrefixOf` s) [0..9]
numbers = ["zero", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine"]
